### ISSUE TITLE
How to make overlay recognized as linked class

### ISSUE THREAD

**Issue by SamHall** [openseadragon/openseadragon#2263]
Demo: [https://chattanoogahistory.com/HTML5-boilerplate/clicktest.php](https://chattanoogahistory.com/HTML5-boilerplate/clicktest.php)

Using [OpenSeadragonHTMLelements.js](https://github.com/antonioxdias/OpenSeadragonHTMLelements) to position SVG(s) over specific locations in canvas.
I use classes to link/open other .dzi's 
This works well for other elements, such as the button labeled 'INDEX'.
However, the red SVG has the same class assigned, but does not react to a click. 
I suspect a JavaScript issue, but cannot figure it out. 

Any assistance would be greatly appreciated. 

Code that links and opens other dzi:
```
// 16
$(document).on('click', '.loc16', function() {
	viewer.open([{
		tileSource: loc16
	}]);
	$("#div1").text("Plate 16");
	setTimeout(function() {
		viewer.viewport.fitBounds(box1);
	}, 400);
});
```

Overlay that is assigned the same class, but not clickable:
```
	var hEl = viewer.HTMLelements()	 
	var img = document.createElement('img')
    img.src = 'https://chattanoogahistory.com/svg/16.svg';	
	img.style.backgroundColor = 'rgba(255,0,0,0.2)'
	img.style.borderRadius = '50%'
	img.className = "loc16";

	 hEl.addElement({
            id: "loc16",
            element: img,
            x: 4600,
            y: 3300,
            width: 300,
            height: 300
          })
```

**Comment by Aiosa** [comment]
I did not really find a solution, however, I observed the following:
 - overlay elements are in the DOM tree placed after the canvas, buttons before and outside `openseadragon-container`
 - contextmenu events work on the red SVG
 - moving the svg to the `#controls` div made it work 
 
The most probable answer is it has something to do with the DOM structure and applied CSS styles. I am afraid I cannot burn too much time on this. Wish you luck :)

**Comment by SamHall** [comment]
> I did not really find a solution, however, I observed the following:
> 
> * overlay elements are in the DOM tree placed after the canvas, buttons before and outside `openseadragon-container`
> * contextmenu events work on the red SVG
> * moving the svg to the `#controls` div made it work
> 
> The most probable answer is it has something to do with the DOM structure and applied CSS styles. I am afraid I cannot burn too much time on this. OSD might be consuming the primary button events as well. Wish you luck :)

Thanks for taking the time to look at it.
You're correct in that the answer having to do with event delegation - I just haven't found the right way to call it.

It may work in the control ID because 'loc16' class is already tagged to a button there.

I'll play around with it and see if I can figure it out.

Thanks.

**Comment by Aiosa** [comment]
Not exactly, I tried `onclick` event manual insertion in html. The above was tested with it. Moving the element in DOM helped so the primary pointer event does not reach your SVG for some reason.

**Comment by SamHall** [comment]
I suppose it's all in the event delegation / bubbling. 
I fight with JavaScript & jQuery more than anything else, including PHP.

**Comment by pearcetm** [comment]
Instead of using jQuery delegated events, try using a `MouseTracker` so your events aren't eaten by the viewer's event handling:

```
new OpenSeadragon.MouseTracker({
    element: yourElement,
    clickHandler: handlerFunc,
    ...
});
```

This should at least get you to the point where you are receiving the events. It will require a little refactoring of where you define the handler(s) but hopefully that won't be too hard.

**Comment by Aiosa** [comment]
I just came by chance to a similar issue, having to listen to events on the viewport overlay. Yes, events are indeed consumed by OSD tracking system. I was a bit skeptic because the events (at least I thought) started at inner DOM nodes and bubbled up unless stopped, since overlays are children or defined after OSD I thought that the event system should not prevent these but it is somehow set so that it does. It has probably something to do with `pointer-events`. This could be mentioned in docs examples about overlays...

**Comment by pearcetm** [comment]
@Aiosa I don't remember exactly where I found info on this originally (whether it was somewhere in the documentation, source code, or just debugging my own problems with this same issue), but I believe that OSD is eating the events during the `capture` phase of the event (when it's propagating from the document down to the deepest nested levels), before the `bubbling` phase where we're used to dealing with them. See this discussion for more info: https://stackoverflow.com/questions/4616694/what-is-event-bubbling-and-capturing

**Comment by Aiosa** [comment]
@pearcetm Aha! Anyway, I updated the docs so that we have it documented. Thank you.

**Comment by SamHall** [comment]
> Instead of using jQuery delegated events, try using a `MouseTracker` so your events aren't eaten by the viewer's event handling:
> 
> ```
> new OpenSeadragon.MouseTracker({
>     element: yourElement,
>     clickHandler: handlerFunc,
>     ...
> });
> ```
> 
> This should at least get you to the point where you are receiving the events. It will require a little refactoring of where you define the handler(s) but hopefully that won't be too hard.

I suspected that was the solution, however, there aren't many complete working examples of MouseTracker that are up to date. As I recall, many changes were made since 3.0? Not sure...  
For someone like me who gets easily lost without complete examples I find myself without a solution.

I tried this but errors abound...

```
new OpenSeadragon.MouseTracker({
    element: 'loc16',
    clickHandler: onCanvasClick, function() {
	viewer.open([{
		tileSource: loc16
	}]);

    }
});
```

**Comment by pearcetm** [comment]
Try this:

```
var hEl = viewer.HTMLelements();	 
var img = document.createElement('img');
img.src = 'https://chattanoogahistory.com/svg/16.svg';	
img.style.backgroundColor = 'rgba(255,0,0,0.2)';
img.style.borderRadius = '50%'
img.className = "loc16";

hEl.addElement({
    id: "loc16",
    element: img,
    x: 4600,
    y: 3300,
    width: 300,
    height: 300
});


new OpenSeadragon.MouseTracker({
    element: img,
    clickHandler: function() {
	viewer.open([{
		tileSource: loc16
	}]);
    }
});
```

**Comment by SamHall** [comment]
OMG - that's perfect and it works as intended. Thank you!
You've made my day!

**Comment by msalsbery** [comment]
Reopening this...there's some confusion and incorrect assumptions regarding the viewer's event handling. 

At the very least this stuff needs better documentation, something that's been on my todo list forever. A lot of changes were made for 3.0 to make this easier...I want to document it

**Comment by iangilman** [comment]
@msalsbery Great! Would be lovely to to clear up confusion :-)

**Comment by msalsbery** [comment]
> > Instead of using jQuery delegated events, try using a `MouseTracker` so your events aren't eaten by the viewer's event handling:
> > ```
> > new OpenSeadragon.MouseTracker({
> >     element: yourElement,
> >     clickHandler: handlerFunc,
> >     ...
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This should at least get you to the point where you are receiving the events. It will require a little refactoring of where you define the handler(s) but hopefully that won't be too hard.
> 
> I suspected that was the solution, however, there aren't many complete working examples of MouseTracker that are up to date. As I recall, many changes were made since 3.0? Not sure... For someone like me who gets easily lost without complete examples I find myself without a solution.
> 
> I tried this but errors abound...
> 
> ```
> new OpenSeadragon.MouseTracker({
>     element: 'loc16',
>     clickHandler: onCanvasClick, function() {
> 	viewer.open([{
> 		tileSource: loc16
> 	}]);
> 
>     }
> });
> ```

@SamHall  I know this issue is old, but FYI and for others that may find their way here: [Here's an up-to-date example using MouseTracker on overlays](https://codepen.io/msalsbery/pen/dyOeXqO)

**Comment by SamHall** [comment]
Much appreciate the updated example!
